prov/sm2/src/sm2_atomic.c

```diff
@@ -0,0 +1,548 @@
+				      const struct fi_rma_ioc *rma_ioc,
+				      size_t ioc_count)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+	atomic_entry->atomic_hdr.rma_ioc_count = ioc_count;
+	memcpy(atomic_entry->atomic_hdr.rma_ioc, rma_ioc,
+	       sizeof(*rma_ioc) * ioc_count);
+}
+
+static void sm2_format_atomic_result(struct sm2_xfer_entry *xfer_entry,
+				     const struct iovec *resultv,
+				     size_t result_count)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+
+	atomic_entry->atomic_hdr.result_iov_count = result_count;
+	memcpy(atomic_entry->atomic_hdr.result_iov, resultv,
+	       sizeof(*resultv) * result_count);
+}
+
+static void sm2_format_atomic_data(struct sm2_xfer_entry *xfer_entry,
+				   const struct iovec *iov, size_t iov_count,
+				   const struct iovec *compare_iov,
+				   size_t compare_count)
+{
+	struct sm2_atomic_entry *atomic_entry =
+		(struct sm2_atomic_entry *) xfer_entry->user_data;
+	size_t comp_size;
+
+	switch (xfer_entry->hdr.op) {
+	case ofi_op_atomic:
+		xfer_entry->hdr.size = ofi_copy_from_iov(
+			atomic_entry->atomic_data.data, SM2_ATOMIC_INJECT_SIZE,
+			iov, iov_count, 0);
+		break;
+	case ofi_op_atomic_fetch:
+		if (xfer_entry->hdr.op != FI_ATOMIC_READ)
+			xfer_entry->hdr.size = ofi_copy_from_iov(
+				atomic_entry->atomic_data.data,
+				SM2_ATOMIC_INJECT_SIZE, iov, iov_count, 0);
+		break;
+	case ofi_op_atomic_compare:
+		xfer_entry->hdr.size = ofi_copy_from_iov(
+			atomic_entry->atomic_data.buf,
+			SM2_ATOMIC_COMP_INJECT_SIZE, iov, iov_count, 0);
+		comp_size = ofi_copy_from_iov(atomic_entry->atomic_data.comp +
+						      xfer_entry->hdr.size,
+					      SM2_ATOMIC_COMP_INJECT_SIZE,
+					      compare_iov, compare_count, 0);
+		if (comp_size != xfer_entry->hdr.size)
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"atomic and compare buffer size mismatch\n");
+		break;
+	default:
+		assert(0);
+		break;
+	}
+}
+
+static ssize_t
+sm2_do_atomic_inject(struct sm2_ep *ep, int64_t peer_gid, uint32_t op,
+		     uint64_t op_flags, uint8_t datatype, uint8_t atomic_op,
+		     const struct iovec *iov, size_t iov_count,
+		     const struct fi_rma_ioc *rma_ioc, size_t rma_ioc_count,
+		     const struct iovec *compare_iov, size_t compare_count,
+		     const struct iovec *result_iov, size_t result_count,
+		     size_t total_len, void *context)
+{
+	struct sm2_xfer_entry *xfer_entry;
+	size_t ret;
+
+	ret = sm2_pop_xfer_entry(ep, &xfer_entry);
+	if (ret)
+		return ret;
+
+	xfer_entry->hdr.proto = sm2_proto_inject;
+
+	sm2_generic_format(xfer_entry, ep->gid, op, 0, 0, op_flags, context);
+	sm2_format_atomic_op(xfer_entry, datatype, atomic_op);
+	sm2_format_atomic_rma_ioc(xfer_entry, rma_ioc, rma_ioc_count);
+	sm2_format_atomic_result(xfer_entry, result_iov, result_count);
+	sm2_format_atomic_data(xfer_entry, iov, iov_count, compare_iov,
+			       compare_count);
```

Comment:
Two things:
- it seems silly to have all the atomic format functions be separate since they all set like one or two things each and do the same casting of the xfer_entry->user_data. Could we move it to one function? Or just do it all here without the need for functions?
- I think it might make more sense to define the user data as a union so we don't have to keep casting it as the atomic entry.

Reply:
> it seems silly to have all the atomic format functions be separate since they all set like one or two things each and do the same casting of the xfer_entry->user_data. Could we move it to one function? Or just do it all here without the need for functions?

Not in love with this but done

> I think it might make more sense to define the user data as a union so we don't have to keep casting it as the atomic entry

I prefer the casts, and we have 3 PR's coming in that are all using casts... I think unions clutter up the structures.